gravity.sh

```diff
@@ -519,25 +531,46 @@ gravity_DownloadBlocklists() {
   gravity_Blackbody=true
 }
 
-# num_total_imported_domains increases for each list processed
-num_total_imported_domains=0
-num_domains=0
-num_non_domains=0
+
+# global variable to indicate if we found ABP style domains during the gravity run
+# is saved in gravtiy's info table to signal FTL if such domains are available
+abp_domains=0
 parseList() {
-  local adlistID="${1}" src="${2}" target="${3}" non_domains sample_non_domains
-  # This sed does the following things:
-  # 1. Remove all lines containing no domains
-  # 2. Remove all domains containing invalid characters. Valid are: a-z, A-Z, 0-9, dot (.), minus (-), underscore (_)
-  # 3. Append ,adlistID to every line
-  # 4. Remove trailing period (see https://github.com/pi-hole/pi-hole/issues/4701)
-  # 5. Ensures there is a newline on the last line
-  sed -r  "/([^\.]+\.)+[^\.]{2,}/!d;/[^a-zA-Z0-9.\_-]/d;s/\.$//;s/$/,${adlistID}/;/.$/a\\" "${src}" >> "${target}"
+  local adlistID="${1}" src="${2}" target="${3}" temp_file temp_file_base non_domains sample_non_domains
+
+  # Create a temporary file for the sed magic instead of using "${target}" directly
+  # this allows to split the sed commands to improve readability
+  # we use a file handle here and remove the temporary file immediately so the content will be deleted in any case
+  # when the script stops
+  temp_file_base="$(mktemp -p "/tmp" --suffix=".gravity")"
+  exec 3>"$temp_file_base"
+  rm "${temp_file_base}"
+  temp_file="/proc/$$/fd/3"
+
+
+  # 1. Add all valid domains (adapted from https://stackoverflow.com/a/30007882)
```

Comment:
Can we make this pattern reusable to prevent possible issues in future when tweaking/changing things?

Something like: 

```suggestion
  # 1. Add all valid domains (adapted from https://stackoverflow.com/a/30007882)
  # no need to include uppercase letters, as we convert to lowercase in gravity_ParseFileIntoDomains() already

  valid_domain_pattern="([a-z0-9]([a-z0-9_-]{0,61}[a-z0-9]){0,1}\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]"
  standard="^${valid_domain_pattern}$"
  abp="^\|\|${valid_domain_pattern}\^$"

  sed -r "/${standard}/!d" "${src}" > "${temp_file}"

  # if there is at least one ABP style domain
  if  grep -E "${abp}" -m 1 -q "${src}"; then
    echo "  ${INFO} List contained AdBlock Plus style domains"
    abp_domains=1
    # 2. Add all supported ABP style lines (||subdomain.domain.tlp^)
    sed -r "/${abp}/!d" "${src}" >> "${temp_file}"
  fi
```

Reply:
Made the suggestion against the wrong place - but the sentiment should be the same!

Reply:
Done.